chemicalx/models/ssiddi.py

```diff
@@ -1,11 +1,318 @@
 """An implementation of the SSI-DDI model."""
 
 from .base import UnimplementedModel
+import torch
+from torch import nn
+from torch.nn import Parameter, Linear
+from torch.nn.modules.container import ModuleList
+from torch import Tensor
+
+from torchdrug.data import PackedGraph
+from torchdrug.layers import GraphAttentionConv
+
+from torch_scatter import scatter, scatter_add, scatter_max
+from chemicalx.constants import TORCHDRUG_NODE_FEATURES
+from chemicalx.data import DrugPairBatch
+from chemicalx.models import Model
+import torch.nn.functional as F
+
 
 __all__ = [
     "SSIDDI",
 ]
 
+def filter_adj(edge_index, edge_attr, perm, num_nodes):
+    mask = perm.new_full((num_nodes, ), -1)
+    i = torch.arange(perm.size(0), dtype=torch.long, device=perm.device)
+    mask[perm] = i
+
+    row, col = edge_index
+    row, col = mask[row], mask[col]
+    mask = (row >= 0) & (col >= 0)
+    row, col = row[mask], col[mask]
+
+    if edge_attr is not None:
+        edge_attr = edge_attr[mask]
+
+    return torch.stack([row, col], dim=0), edge_attr
+
+
+def topk(x, ratio, batch, min_score=None, tol=1e-7):
+    if min_score is not None:
+        # Make sure that we do not drop all nodes in a graph.
+        scores_max = scatter_max(x, batch)[0].index_select(0, batch) - tol
+        scores_min = scores_max.clamp(max=min_score)
+
+        perm = (x > scores_min).nonzero(as_tuple=False).view(-1)
+    else:
+        num_nodes = scatter_add(batch.new_ones(x.size(0)), batch, dim=0)
+        batch_size, max_num_nodes = num_nodes.size(0), num_nodes.max().item()
+
+        cum_num_nodes = torch.cat(
+            [num_nodes.new_zeros(1),
+             num_nodes.cumsum(dim=0)[:-1]], dim=0)
+
+        index = torch.arange(batch.size(0), dtype=torch.long, device=x.device)
+        index = (index - cum_num_nodes[batch]) + (batch * max_num_nodes)
+
+        dense_x = x.new_full((batch_size * max_num_nodes, ),
+                             torch.finfo(x.dtype).min)
+        dense_x[index] = x
+        dense_x = dense_x.view(batch_size, max_num_nodes)
+
+        _, perm = dense_x.sort(dim=-1, descending=True)
+
+        perm = perm + cum_num_nodes.view(-1, 1)
+        perm = perm.view(-1)
+
+        if isinstance(ratio, int):
+            k = num_nodes.new_full((num_nodes.size(0), ), ratio)
+            k = torch.min(k, num_nodes)
+        else:
+            k = (ratio * num_nodes.to(torch.float)).ceil().to(torch.long)
+
+        mask = [
+            torch.arange(k[i], dtype=torch.long, device=x.device) +
+            i * max_num_nodes for i in range(batch_size)
+        ]
+        mask = torch.cat(mask, dim=0)
+
+        perm = perm[mask]
+
+    return perm
+
+
+def softmax(src, batch):
```

Comment:
is this different from https://pytorch-scatter.readthedocs.io/en/1.4.0/composite/softmax.html#torch_scatter.composite.scatter_softmax?

Reply:
No, they are the same. I think we can directly use this package here. Thanks for the check.